internal/uvm/scsi/manager.go

```diff
@@ -0,0 +1,170 @@
+)
+
+// Manager is the primary entrypoint for managing SCSI devices on a VM.
+// It tracks the state of what devices have been attached to the VM, and
+// mounted inside the guest OS.
+type Manager struct {
+	attachManager *attachManager
+	mountManager  *mountManager
+}
+
+// Slot represents a single SCSI slot, consisting of a controller and LUN.
+type Slot struct {
+	Controller uint
+	LUN        uint
+}
+
+// NewManager creates a new Manager using the provided backend [Attacher],
+// [Mounter], and [Unplugger], as well as other configuration parameters.
+//
+// guestMountFmt is the format string to use for mounts of SCSI devices in
+// the guest OS. It should have a single %d format parameter.
+//
+// reservedSlots indicates which SCSI slots to treat as already used. They
+// will not be handed out again by the Manager.
+func NewManager(
+	attacher Attacher,
+	mounter Mounter,
+	unplugger Unplugger,
+	numControllers int,
+	numLUNsPerController int,
+	guestMountFmt string,
+	reservedSlots []Slot,
+) *Manager {
+	am := newAttachManager(attacher, unplugger, numControllers, numLUNsPerController, reservedSlots)
+	mm := newMountManager(mounter, guestMountFmt)
+	return &Manager{am, mm}
+}
+
+// Mount represents a SCSI device that has been attached to a VM, and potentially
+// also mounted into the guest OS.
+type Mount struct {
+	mgr         *Manager
+	controller  uint
+	lun         uint
+	guestPath   string
+	releaseOnce sync.Once
+}
+
+// Controller returns the controller number that the SCSI device is attached to.
+func (m *Mount) Controller() uint {
+	return m.controller
+}
+
+// LUN returns the LUN number that the SCSI device is attached to.
+func (m *Mount) LUN() uint {
+	return m.lun
+}
+
+// GuestPath returns the path inside the guest OS where the SCSI device was mounted.
+// Will return an empty string if no guest mount was performed.
+func (m *Mount) GuestPath() string {
+	return m.guestPath
+}
+
+// Release releases the SCSI mount. Refcount tracking is used in case multiple instances
+// of the same attachment or mount are used. If the refcount for the guest OS mount
+// reaches 0, the guest OS mount is removed. If the refcount for the SCSI attachment
+// reaches 0, the SCSI attachment is removed.
+func (m *Mount) Release(ctx context.Context) (err error) {
+	err = ErrAlreadyReleased
+	m.releaseOnce.Do(func() {
+		err = m.mgr.remove(ctx, m.controller, m.lun, m.guestPath)
+	})
+	return
+}
+
+func (m *Manager) Add(ctx context.Context, attachConfig *AttachConfig, mountConfig *MountConfig) (_ *Mount, err error) {
+	if m == nil {
+		return nil, ErrNotInitialized
+	}
+	controller, lun, err := m.attachManager.attach(ctx, attachConfig)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			_, _ = m.attachManager.detach(ctx, controller, lun)
+		}
+	}()
+
+	var guestPath string
+	if mountConfig != nil {
+		guestPath, err = m.mountManager.mount(ctx, controller, lun, mountConfig)
```

Comment:
should we check that the readonly flags match? since trying to mount as readwrite a scsi that was attached as readonly will likely cause weird errors down that line that we could have cleanly caught and warned about here

Reply:
I think this is a good argument for having a higher level interface over "storage". :) Let's discuss on the [existing thread](https://github.com/microsoft/hcsshim/pull/1744#discussion_r1178501433) for that.

Reply:
I've switched back to a simpler client-facing interface, where only a single read-only value is provided, rather than separately for attach/mount.